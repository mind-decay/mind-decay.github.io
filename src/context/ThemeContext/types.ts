import { Nullable } from 'types';

export type ThemeType = 'light' | 'dark';

export type ThemeContextType = {
  theme: ThemeType;
  toggleTheme: () => void;
};

export const isThemeType = (theme: Nullable<string>): theme is ThemeType => {
  return typeof theme === 'string' && ['light', 'dark'].includes(theme);
};
