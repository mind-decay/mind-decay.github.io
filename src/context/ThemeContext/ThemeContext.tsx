import { ReactNode, createContext, useContext, useEffect, useState } from 'react';
import { ThemeProvider as StyledThemeProvider } from 'styled-components';

import { ThemeContextType, ThemeType, isThemeType } from './types';
import { Nullable } from 'types/helpers';

import { darkTheme, lightTheme } from 'styles/themes';

const ThemeContext = createContext<Nullable<ThemeContextType>>(null);

export const useTheme = (): ThemeContextType => {
  const context = useContext(ThemeContext);

  if (!context) {
    throw new Error('useTheme must be used within a ThemeProvider');
  }

  return context;
};

type ThemeProviderProps = {
  children: ReactNode;
};

export const ThemeProvider = ({ children }: ThemeProviderProps) => {
  const [theme, setTheme] = useState<ThemeType>('dark');
  const isLightTheme = theme === 'light';

  const toggleTheme = () => {
    const updatedTheme = isLightTheme ? 'dark' : 'light';
    setTheme(updatedTheme);
    localStorage.setItem('theme', updatedTheme);
  };

  useEffect(() => {
    const savedTheme = localStorage.getItem('theme');
    const prefersLight = window.matchMedia && window.matchMedia('(prefers-color-scheme: light)').matches;
    const checkCurrentTheme = isThemeType(savedTheme);

    if (checkCurrentTheme) {
      setTheme(savedTheme);
    } else if (prefersLight) {
      setTheme('light');
    }
  }, []);

  return (
    <ThemeContext.Provider value={{ theme, toggleTheme }}>
      <StyledThemeProvider theme={isLightTheme ? lightTheme : darkTheme}>{children}</StyledThemeProvider>
    </ThemeContext.Provider>
  );
};
