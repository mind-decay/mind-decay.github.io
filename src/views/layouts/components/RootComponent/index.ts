export * from './RootComponent';
export * from './routes';
