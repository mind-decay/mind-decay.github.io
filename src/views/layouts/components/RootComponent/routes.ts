export const mainRoutes = {
  Root: '/',
} as const;
