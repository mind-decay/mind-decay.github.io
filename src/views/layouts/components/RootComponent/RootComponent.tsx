import { Suspense, lazy } from 'react';
import { Route, Routes } from 'react-router-dom';

import { Loader } from 'views/components/Loader';
import { Wrapper } from './components/Wrapper';
const MainLayout = lazy(() => import('views/layouts/MainLayout/MainLayout'));
const Home = lazy(() => import('views/pages/Home/Home'));

import { mainRoutes } from './routes';

export const RootComponent = () => {
  return (
    <Suspense fallback={<Loader />}>
      <Wrapper>
        <Routes>
          <Route element={<MainLayout />}>
            <Route path={mainRoutes.Root} element={<Home />} />
          </Route>
        </Routes>
      </Wrapper>
    </Suspense>
  );
};
