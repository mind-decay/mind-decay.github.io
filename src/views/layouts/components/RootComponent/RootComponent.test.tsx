import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import { BrowserRouter } from 'react-router-dom';

import { ThemeProvider } from 'context/ThemeContext/ThemeContext';
import { RootComponent } from './RootComponent';

test('renders root component', () => {
  const { container } = render(
    <BrowserRouter>
      <ThemeProvider>
        <RootComponent />
      </ThemeProvider>
    </BrowserRouter>,
  );

  expect(container).toMatchSnapshot();
});
