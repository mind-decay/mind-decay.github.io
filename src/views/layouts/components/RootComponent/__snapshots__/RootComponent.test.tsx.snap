// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders root component 1`] = `
<div>
  <div
    class="sc-beySPh iohzhV"
  >
    <div
      class="sc-guDLey lagWuR"
      data-testid="loader"
    />
  </div>
</div>
`;
